src/outputs/rcp.c

```diff
@@ -0,0 +1,1484 @@
+ *
+ * RCP spec page 10 - Protocol Summary
+ *   RCP was designed with simplicity and completeness as primary requirements.
+ *   Commands and results are exchanged as short transmissions across a high-
+ *   speed interface like a serial port, telnet connection, or parallel interface. Each
+ *   command is composed of a short ASCII command id string, generally just zero or
+ *   one parameters, and the two-byte terminator CRLF. All command results from
+ *   the RCP host are composed of the command-id of the client command that
+ *   caused this result followed by a result string and the two-byte CRLF terminator
+ *
+ *   ...
+ *
+ *   RCP commands can be loosely categorized by the way in which they execute:
+ *   synchronous commands, transacted commands, and subscription commands.
+ *   Synchronous commands return their results immediately, and do not block the
+ *   host device during execution. Transacted commands are commands that might
+ *   require a long extent of time to complete, some as long as ten seconds or more!
+ *   These commands generally depend on sending and receiving data from the
+ *   network, which is why their completion time is non-deterministic. Transacted
+ *   commands run asynchronously, or “in the background,” in an RCP session, and
+ *   allow the client to issue other commands or cancel the command while it is in
+ *   process.
+ *
+ * Communcations within this module only use the sync subset of commands from
+ * the RCP spec
+ */
+enum rcp_state
+{
+  RCP_STATE_SETUP,
+  RCP_STATE_SETUP_WAKEUP,
+
+  RCP_STATE_SETUP_GET_CONNECTED_SERVER,
+  RCP_STATE_SETUP_SERVER_DISCONNECT_TRANS_INIT,
+  RCP_STATE_SETUP_SERVER_DISCONNECT_DISCONNECTED,
+  RCP_STATE_SETUP_SERVER_DISCONNECT_TRANS_END,
+  RCP_STATE_SETUP_SERVER_DISCONNECT,
+  RCP_STATE_SETUP_SET_SERVER_FILTER,
+  RCP_STATE_SETUP_LIST_SERVERS_RESULT_SIZE,
+  RCP_STATE_SETUP_LIST_SERVERS_INTERNET_RADIO,
+  RCP_STATE_SETUP_LIST_SERVERS_RESULTS_END,
+  RCP_STATE_SETUP_LIST_SERVERS,
+  RCP_STATE_SETUP_SERVER_CONNECT_TRANS_INIT,
+  RCP_STATE_SETUP_SERVER_CONNECT_CONNECTED,
+  RCP_STATE_SETUP_SERVER_CONNECT_TRANS_END,
+  RCP_STATE_SETUP_SERVER_CONNECT,
+
+  RCP_STATE_SETUP_VOL_GET,
+
+  RCP_STATE_QUEUING_CLEAR,
+  RCP_STATE_QUEUING_SET_TITLE,
+  RCP_STATE_QUEUING_SET_PLAYLIST_URL,
+  RCP_STATE_QUEUING_SET_REMOTE_STREAM,
+  RCP_STATE_QUEUING_PLAY,
+
+  RCP_STATE_STREAMING,
+
+  RCP_STATE_VOL_GET,
+  RCP_STATE_VOL_SET,
+
+  RCP_STATE_STOPPING,
+
+  RCP_STATE_SHUTDOWN_STOPPED,
+  RCP_STATE_SHUTDOWN_GET_CONNECTED_SERVER,
+  RCP_STATE_SHUTDOWN_SERVER_DISCONNECT_TRANS_INIT,
+  RCP_STATE_SHUTDOWN_SERVER_DISCONNECT_DISCONNECTED,
+  RCP_STATE_SHUTDOWN_SERVER_DISCONNECT_TRANS_END,
+  RCP_STATE_SHUTDOWN_SERVER_DISCONNECT,
+
+  // grouped order
+  RCP_STATE_STOPPED,
+  RCP_STATE_STANDBY,
+  RCP_STATE_DISCONNECTED,
+  RCP_STATE_FAILED,
+
+  RCP_STATE_MAX
+};
+
+struct rcp_state_map
+{
+  enum rcp_state state;
+  char *cmd;
+  bool has_arg;
+};
+
+static char *rcp_cmd_title;
```

Comment:
Is a global really required?

Reply:
we can drop the global - you prefer each session to have a copy of `cfg_getstr(cfg_getsec(cfg, "library"), "name");` or to call this `cfg_...()` when required?

Reply:
dropped global, calling the `cfg_..()` at the point we need (on session startup/handshake with remote)